scylla-cdc/src/reader.rs

```diff
@@ -0,0 +1,333 @@
+}
+
+impl StreamID {
+    pub fn new(stream_id: Vec<u8>) -> StreamID {
+        StreamID { id: stream_id }
+    }
+}
+
+pub struct StreamReader {
+    session: Arc<Session>,
+    stream_id_vec: Vec<StreamID>,
+    lower_timestamp: chrono::Duration,
+    window_size: chrono::Duration,
+    safety_interval: chrono::Duration,
+    upper_timestamp: tokio::sync::Mutex<Option<chrono::Duration>>,
+    sleep_interval: time::Duration,
+}
+
+impl StreamReader {
+    pub fn new(
+        session: &Arc<Session>,
+        stream_ids: Vec<StreamID>,
+        start_timestamp: chrono::Duration,
+        window_size: chrono::Duration,
+        safety_interval: chrono::Duration,
+        sleep_interval: time::Duration,
+    ) -> StreamReader {
+        StreamReader {
+            session: session.clone(),
+            stream_id_vec: stream_ids,
+            lower_timestamp: start_timestamp,
+            window_size,
+            safety_interval,
+            upper_timestamp: Default::default(),
+            sleep_interval,
+        }
+    }
+
+    pub async fn set_upper_timestamp(&mut self, new_upper_timestamp: chrono::Duration) {
+        let mut guard = self.upper_timestamp.lock().await;
+        *guard = Some(new_upper_timestamp);
+    }
+
+    pub async fn fetch_cdc(
+        &mut self,
+        keyspace: String,
+        table_name: String,
+    ) -> Result<(), Box<dyn Error>> {
+        let query = format!(
+            "SELECT pk,t,v,s FROM {}.{}_scylla_cdc_log \
+            WHERE \"cdc$stream_id\" in (?) \
+            AND \"cdc$time\" >= maxTimeuuid(?) \
+            AND \"cdc$time\" < minTimeuuid(?)",
+            keyspace, table_name
+        );
+        let query_base = self.session.prepare(query).await?;
+        let mut lower_timestamp = self.lower_timestamp;
+
+        loop {
+            let now = chrono::Duration::from_std(
+                SystemTime::now().duration_since(SystemTime::UNIX_EPOCH)?,
+            )?;
+
+            let guard = self.upper_timestamp.lock().await;
+            let upper_timestamp = guard.unwrap_or(max(
+                lower_timestamp,
+                min(
+                    lower_timestamp + self.window_size,
+                    now - self.safety_interval,
+                ),
+            ));
+
+            let mut rows_stream = self
+                .session
+                .execute_iter(
+                    query_base.clone(),
+                    (
+                        &self.stream_id_vec,
+                        Timestamp(lower_timestamp),
+                        Timestamp(upper_timestamp),
+                    ),
+                )
+                .await?
+                .into_typed::<(i32, i32, String, String)>();
+
+            while let Some(row) = rows_stream.next().await {
+                let (pk, t, v, s) = row?;
+                println!("pk: {} t: {} v: {} s: {}", pk, t, v, s);
+            }
+
+            if lower_timestamp == upper_timestamp {
+                break;
+            }
```

Comment:
I think you should compare with `self.upper_timestamp` here, not the local variable `upper_timestamp`. Moreover, for safety, the comparison should use `>=` instead of the equality operator.

Reply:
As I understand, we need either the preset `upper_timestamp` or the "safe" `upper_timestamp` calculated at the moment, could you please clarify why `self.upper_timestamp` should be in the comparison.

Reply:
The local variables `lower_timestamp` and `upper_timestamp` mean the range of the time window you are reading from in this iteration. The `self.upper_timestamp` means the timestamp at which we want the stream reader to stop. When the stream reader detects that it read all changes up to `self.upper_timestamp`, we want the `fetch_cdc` function to stop.

By the way, I'd suggest renaming `lower_timestamp` -> `window_begin` and `upper_timestamp` (the local variable) -> `window_end` in order to reduce confusion.